sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkSpan.java

```diff
@@ -493,12 +498,16 @@ public void end(long timestamp, TimeUnit unit) {
 
   private void endInternal(long endEpochNanos) {
     synchronized (lock) {
-      if (hasEnded) {
-        logger.log(Level.FINE, "Calling end() on an ended Span.");
+      if (hasEnded != EndState.NOT_ENDED) {
+        logger.log(Level.FINE, "Calling end() on an ended or ending Span.");
         return;
       }
+      hasEnded = EndState.ENDING;
       this.endEpochNanos = endEpochNanos;
-      hasEnded = true;
+      if (spanProcessor.isOnEndingRequired()) {
+        spanProcessor.onEnding(this);
```

Comment:
I think it's probably a mistake to be calling this under the lock. I believe we only want to do internal state management while the lock is being held, not call not-under-our-control methods provided by users.

Reply:
The lock was actually intentionally hold to cover this part of the spec:
> The SDK MUST guarantee that the span can no longer be modified by any other thread
before invoking `OnEnding` of the first `SpanProcessor`. From that point on, modifications
are only allowed synchronously from within the invoked `OnEnding` callbacks.

This part was added to the spec so that SpanProcessors can be sure that the state they see the span in is actually consistent and not prone to race conditions due to span modifications from the application code.

However, I now switched to a different way of implementing this, because as you said exposing the lock is probably not the best way of doing it here.